### ISSUE TITLE
Non-deterministic graph when exporting to ONNX even when random seed is set

### ISSUE THREAD

**Issue by danibene** [onnx/sklearn-onnx#1001]
Hello,

When exporting a model multiple times trained with the same configuration, data, and random seed, I noticed that the ONNX files, and their corresponding graphs, differed. Here is some [sample code](https://gist.github.com/danibene/74a8986a20acae78d883c893878b6366) demonstrating the issue:

```python
from pathlib import Path

import numpy as np
import onnx
from skl2onnx import convert_sklearn
from skl2onnx.common.data_types import FloatTensorType
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier

RANDOM_SEED = 42
TARGET_OPSET = 12
EXPORT_PARENT_PATH = Path("exported_to_onnx")


def export_sample_onnx(iteration=1):
    # Load the dataset
    iris = load_iris()
    X = iris.data
    y = iris.target
    clf = RandomForestClassifier(random_state=RANDOM_SEED)
    clf.fit(X, y)
    initial_type = [("float_input", FloatTensorType([1, len(X)]))]
    converted = convert_sklearn(clf, initial_types=initial_type, target_opset=TARGET_OPSET)
    export_path = Path(EXPORT_PARENT_PATH, "model_" + str(iteration) + ".onnx")
    Path(export_path).parent.mkdir(parents=True, exist_ok=True)
    with open(export_path, "wb") as f:
        f.write(converted.SerializeToString())


def compare_onnx_graphs(model_1_onnx_path: Path, model_2_onnx_path: Path):
    # Load the ONNX models
    model_1 = onnx.load(str(model_1_onnx_path))
    model_2 = onnx.load(str(model_2_onnx_path))

    # Get the graphs from the models
    graph_1 = model_1.graph
    graph_2 = model_2.graph

    diff_graphs = []

    # Compare the number of nodes
    if len(graph_1.node) != len(graph_2.node):
        print("The number of nodes in the graphs is different.")
        return [{"graph_1": node, "graph_2": "not_same_len"} for node in graph_1.node]

    # Compare each node in the graphs
    for node_1, node_2 in zip(graph_1.node, graph_2.node):
        if not _check_same_onnx_nodes(node_1, node_2):
            diff_graphs.append({"graph_1": node_1, "graph_2": node_2})

    return diff_graphs


def _check_same_onnx_nodes(node_1: onnx.NodeProto, node_2: onnx.NodeProto):
    if node_1.attribute != node_2.attribute:
        return False
    elif node_1.input != node_2.input:
        return False
    elif node_1.output != node_2.output:
        return False
    elif node_1.op_type != node_2.op_type:
        return False
    else:
        return True


for i in range(5):
    export_sample_onnx(i)

onnx_file_paths = list(EXPORT_PARENT_PATH.rglob("*.onnx"))

for i in range(len(onnx_file_paths) - 1):
    diffs_graphs = compare_onnx_graphs(onnx_file_paths[i], onnx_file_paths[i + 1])
    if len(diffs_graphs) > 0:
        print(diffs_graphs)
```

Which gave me the following output:

```
[{'graph_1': input: "label"
output: "output_label"
name: "Cast"
op_type: "Cast"
attribute {
  name: "to"
  i: 7
  type: INT
}
domain: ""
, 'graph_2': input: "probabilities"
output: "output_probability"
name: "ZipMap"
op_type: "ZipMap"
attribute {
  name: "classlabels_int64s"
  ints: 0
  ints: 1
  ints: 2
  type: INTS
}
domain: "ai.onnx.ml"
}, {'graph_1': input: "probabilities"
output: "output_probability"
name: "ZipMap"
op_type: "ZipMap"
attribute {
  name: "classlabels_int64s"
  ints: 0
  ints: 1
  ints: 2
  type: INTS
}
domain: "ai.onnx.ml"
, 'graph_2': input: "label"
output: "output_label"
name: "Cast"
op_type: "Cast"
attribute {
  name: "to"
  i: 7
  type: INT
}
domain: ""
}]
[{'graph_1': input: "probabilities"
output: "output_probability"
name: "ZipMap"
op_type: "ZipMap"
attribute {
  name: "classlabels_int64s"
  ints: 0
  ints: 1
  ints: 2
  type: INTS
}
domain: "ai.onnx.ml"
, 'graph_2': input: "label"
output: "output_label"
name: "Cast"
op_type: "Cast"
attribute {
  name: "to"
  i: 7
  type: INT
}
domain: ""
}, {'graph_1': input: "label"
output: "output_label"
name: "Cast"
op_type: "Cast"
attribute {
  name: "to"
  i: 7
  type: INT
}
domain: ""
, 'graph_2': input: "probabilities"
output: "output_probability"
name: "ZipMap"
op_type: "ZipMap"
attribute {
  name: "classlabels_int64s"
  ints: 0
  ints: 1
  ints: 2
  type: INTS
}
domain: "ai.onnx.ml"
}]
```

Any help with understanding why this occurs and if it is possible to make the exports reproducible would be appreciated!

**Comment by xadupre** [comment]
PR #1004 was merged. Let me know if it improves in your case. If not, feel free to reopen this issue.

**Comment by danibene** [comment]
Hi @xadupre , thank you for your response and efforts to address this! Unfortunately, it did not seem to improve in my case: https://gist.github.com/danibene/b2d5ec03a6c259246e590cbf1dda4e93
